### ISSUE TITLE
Code for Monocular 3D pose estimation from images

### ISSUE THREAD

**Issue by moonbow721** [hynann/NRDF#4]
Hi there, Thanks for the great work!

Could you please provide the code for 3D pose estimation? I have tried the following code as a prior term and integrated it into fitting frameworks like SMPLify. I have adjusted the weight from 1.0 to 1e5 but it didn't work (results were even worse than no prior term).

```
class NRDF(nn.Module):
    def __init__(self, model_dir, weight=1.0):
        super().__init__()
        from configs.config import load_config
        from lib.model.nrdf import NRDF
        checkpoint_path = os.path.join(model_dir, 'checkpoints', 'checkpoint_epoch_best.tar')
        config_file = os.path.join(model_dir, 'config.yaml')
        self.weight = weight

        self.model = NRDF(load_config(config_file))
        checkpoint = torch.load(checkpoint_path, map_location='cpu')['model_state_dict']
        self.model.load_state_dict(checkpoint)
        self.model.eval()

    def forward(self, pose, *args):
        pose_quat = axis_angle_to_quaternion(pose.view(-1, 21, 3))
        # pose_quat, _ = quat_flip(pose_quat)
        # pose_quat = torch.nn.functional.normalize(pose_quat, dim=-1)
        dist_pred = self.model(pose_quat, train=False)['dist_pred']

        return torch.mean(dist_pred) * self.weight
```

**Comment by hynann** [comment]
Hi Junzhe, thank you for your interest, and sorry for the late reply. We’ve just added the pose estimation code to our repository, but it’s not fully operational due to some dependency issues. We plan to fix these in the coming weeks.

For integrating the NRDF prior into other optimization processes, you can take a look at the `optimize` function in `lib/exp/hps.py`. Please note that we update the gradient in the tangent space rather than directly in Euclidean space.

**Comment by GhostLate** [comment]
Hi, @hynann, thank your for your work!

I also tried to adapt NRDF denoise code to use as a pose prior in the SMPLify like repo and ran into the same problem, I get much better results with VPoser than with NRDF.

Here is my class pose prior NRDF class:

`
class NRDF_Prior(nn.Module):

    def __init__(self, device):
        super().__init__()

        model_dir="./core/losses/pose_priors/nrdf/pretrained_models/amass_softplus_l1_0.0001_10000_dist0.5_eik0.0_man0.1"
        checkpoint_path = os.path.join(model_dir, 'checkpoints', 'checkpoint_epoch_best.tar')
        config_file = os.path.join(model_dir, 'config.yaml')

        self.model = NRDF(config.load_config(config_file)).to(device)
        self.model.load_state_dict(torch.load(checkpoint_path, map_location=device)['model_state_dict'])
        self.model.eval()

    def get_loss(self, pose):
        noisy_pose_quat = axis_angle_to_quaternion(pose.reshape(-1, 21, 3))  # [bs, nj, 4]
        loss = self.model(noisy_pose_quat, train=False)['dist_pred']
        return loss.mean()

    def forward(self, pose):
        assert len(pose.shape) == 2
        return self.get_loss(pose)
`

I also looked at `lib/exp/hps.py`, but you use PoseNDF there. Do you have the same example with NRDF model?
